### PR TITLE
Add support for EIP2098 "short signatures" in the ECDSA library

### LINKED ISSUE #2539: Add support for EIP2098 "short signatures" in the ECDSA library
**🧐 Motivation**
Secp256k1 signatures produced by EOA to sign transaction and messages are usually encode on 65 bytes:
- R: 32 bytes
- S: 32 bytes
- V: 1 bytes.

However, the V parameter can be reduced to only one single bit, and the S parameter only requiers 255 bits to be encoded (due to parity constraints). This result in the signature space fitting inside 64 bytes! [EIP2098](https://eips.ethereum.org/EIPS/eip-2098) describes how to pack the R,S,V values into a shorter R,VS format tat would take less space and thus me less expensive to transmit/store

Ethers.js already include tooling to produce such signatures.

**📝 Details**
EIP details are available [here](https://eips.ethereum.org/EIPS/eip-2098).
Implementation in the ECDSA only requires to accept 65 & 64 bytes long signatures, split them into R,S,V values, and call the recover method. Since the changes that only in the "bytes unpacking" part of the function, and the actual cryptography check done in ECDSA.recover are not modified or bypassed, there should be no security concern. 

```
function recover(bytes32 hash, bytes memory signature) internal pure returns (address) {
    // Divide the signature in r, s and v variables
    bytes32 r;
    bytes32 s;
    uint8 v;

    // Check the signature length
    // - case 65: r,s,v signature (standard)
    // - case 64: r,vs signature (cf https://eips.ethereum.org/EIPS/eip-2098)
    if (signature.length == 65) {
        // ecrecover takes the signature parameters, and the only way to get them
        // currently is to use assembly.
        // solhint-disable-next-line no-inline-assembly
        assembly {
            r := mload(add(signature, 0x20))
            s := mload(add(signature, 0x40))
            v := byte(0, mload(add(signature, 0x60)))
        }
    } else if (signature.length == 64) {
        // ecrecover takes the signature parameters, and the only way to get them
        // currently is to use assembly.
        // solhint-disable-next-line no-inline-assembly
        assembly {
            r := mload(add(signature, 0x20))
            s := and(mload(add(signature, 0x40)), 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
            v := add(shr(7, byte(0, mload(add(signature, 0x40)))), 27)
        }
    } else {
        revert("ECDSA: invalid signature length");
    }

    return recover(hash, v, r, s);
}
```

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #2539

#### PR Checklist
- [x] Tests
- [x] Changelog entry

### CODE REVIEW

**Comment by frangio** [comment]
Waiting for a reply about the EIP becoming Final. https://github.com/ethereum/EIPs/issues/2440

File: test/utils/cryptography/ECDSA.test.js
Comment (frangio) [inline test/utils/cryptography/ECDSA.test.js:15]:
```suggestion
  short[32] |= (long[64] % 27) << 7; // set the first bit of the 32nd byte to the v parity bit
```

File: test/utils/cryptography/ECDSA.test.js
Comment (frangio) [inline test/utils/cryptography/ECDSA.test.js:23]:
```suggestion
  short[32] &= (1 << 7) - 1; // zero out the first bit of 1 the 32nd byte
```

File: contracts/utils/cryptography/ECDSA.sol
Comment (frangio) [inline contracts/utils/cryptography/ECDSA.sol:52]: Cheaper and closer to the sample Python code in EIP-2098.

```suggestion
                v := add(shr(vs, 255), 27)
```

File: test/utils/cryptography/ECDSA.test.js
Comment (frangio) [inline test/utils/cryptography/ECDSA.test.js:81]:
```suggestion
      context('short EIP2098 format', function () {
```

File: test/utils/cryptography/ECDSA.test.js
Comment (frangio) [inline test/utils/cryptography/ECDSA.test.js:122]:
```suggestion
      context('short EIP2098 format', function () {
```

File: test/utils/cryptography/ECDSA.test.js
Comment (frangio) [inline test/utils/cryptography/ECDSA.test.js:82]: A lot of the tests in this file have `it('works'` and `it('reverts'` as the only tests in a describe block. This is really unnecessary, we should be writing the test directly as in: `it('recovers EIP2098 format', ...` and `it('reverts with ...`.

Can you clean this up while we're working on this file?

**Comment by frangio** [comment]
We've decided to move forward with this PR even though the corresponding EIP is a Draft because we don't see how it could change in a backwards incompatible way, and it seems to be on the path to be Finalized.

**Review by frangio — COMMENTED** [review]
Ready to merge after my comments here are addressed.

File: test/utils/cryptography/ECDSA.test.js
Comment (Amxx) [reply test/utils/cryptography/ECDSA.test.js:82]: Lets discuss that in sync. You know I'm bad with names, and its not just about the new tests, its also about the old ones...

Comment (frangio) [reply test/utils/cryptography/ECDSA.test.js:82]: I'm ok with changing to `works with <name of describe>` and `reverts with <name of describe>`. If you don't feel comfortable with it let's move ahead and merge if this is ready and we can talk about the test titles afterwards.

**Comment by frangio** [comment]
There should be some documentation about the format of the `signature` argument. We don't even have documentation for the long format...

**Review by frangio — APPROVED** [review]
Thank you!

**Pull request merged**
